### PR TITLE
Dw 5066

### PR DESCRIPTION
**Pull request by abdu-n**
Added section to generate_analytical_dataset_from_htme.py for saving correlation_id and s3_prefix to adg_params.csv
Added send_sn_message.py for sending ADG sending message
Added section for send_sns_message.py on step.tf
Added step to steps.yaml.tpl for send_snn_message.py
Output adg_completion_status_sns ARN from adg_completion_status_sns.tf

Please note, there are a LOT of changes in the python files, but most of them come from the Black formatter which we all agreed to use.

### CODE REVIEW

File: steps/generate_dataset_from_htme.py [hunk 1]
```diff
@@ -429,6 +431,14 @@ def log_end_of_batch(correlation_id, run_id, status, dynamodb=None):
                          correlation_id, run_id, str(ex))
         sys.exit(-1)
 
+def create_adg_status_csv(correlation_id, s3_prefix, publish_bucket, s3_client):
```
Comment (sandeepg-github) [inline steps/generate_dataset_from_htme.py:434]: We would also need some unit tests for this. For reference I have written some tests in the branch attached to this ticket.

**Review by sandeepg-github — CHANGES_REQUESTED** [review]

**Review by TheDogfather17 — CHANGES_REQUESTED** [review]

File: steps/send_notification.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import json
+import boto3
+import csv
+
+def send_sns_message():
+    payload = {}
+    _sns_client = boto3.client(service_name="sns")
+
+    publish_bucket = "${publish_bucket}"
```
[the remark this answers is not in the document]
Comment (abdu-n) [reply steps/send_notification.py:9]: Tried publish_bucket = os.getenv("S3_PUBLISH_BUCKET"), it did not work, it returning none

**Review by mattburgess — COMMENTED** [review]
On top of what others have said, does this even need to be a separate step?

Does the driver process of the main PySpark script not keep track of whether its overall status is success or failure? Maybe we can get hold of that via a SparkListener (though my casual googling hasn't found a native PySpark interface to SparkListeners :( )

And now that I've asked that question, it occurs to me that this PR currently doesn't output a status at all, so consumers have no idea whether the run was successful or not and therefore whether the `s3_prefix` column points to a location with a partial dataset in it.

**Comment by mattburgess** [comment]
> On top of what others have said, does this even need to be a separate step?
> 
> Does the driver process of the main PySpark script not keep track of whether its overall status is success or failure? Maybe we can get hold of that via a SparkListener (though my casual googling hasn't found a native PySpark interface to SparkListeners :( )
> 
> And now that I've asked that question, it occurs to me that this PR currently doesn't output a status at all, so consumers have no idea whether the run was successful or not and therefore whether the `s3_prefix` column points to a location with a partial dataset in it.

Looks like there's a `status` interface. There's an example at https://github.com/apache/spark/blob/master/examples/src/main/python/status_api_demo.py. API docs are at https://spark.apache.org/docs/latest/api/python/pyspark.html#pyspark.StatusTracker

File: adg_params.csv [hunk 3]
```diff
@@ -0,0 +1,2 @@
+CORRELATION_ID,S3_PREFIX
```
Comment (sandeepg-github) [inline adg_params.csv:1]: this file needs deleted as it doesn't belong to source code

**Review by TheDogfather17 — APPROVED** [review]

**Review by sandeepg-github — APPROVED** [review]

**Pull request merged**